src/core/middlewares/getIpfs.js

```diff
@@ -0,0 +1,77 @@
+'use strict'
+
+const startIpfs = require('../commands/start-ipfs')
+const request = require('ipfs-registry-mirror-common/utils/retry-request')
+const timeout = require('ipfs-registry-mirror-common/utils/timeout-promise')
+
+const cleanUpOps = []
+
+const cleanUp = async () => {
+  Promise.all(
+    cleanUpOps.map(op => op())
+  )
+    .then(() => {
+      process.exit(0)
+    })
+}
+
+process.on('SIGTERM', cleanUp)
+process.on('SIGINT', cleanUp)
+
+const ipfsInstance = (function() {
+  let ipfs = null
+
+  return {
+    get: () => ipfs,
+    set: (instance) => { ipfs = instance }
+  }
+}())
+
+module.exports = (options) => {
+  return async (req, res, next) => {
+    if (ipfsInstance.get() !== null) {
+      res.locals.ipfs = ipfsInstance.get()
+      return next()
+    }
+
+    const ipfs = await startIpfs(options)
```

Comment:
This does not return immediately so multiple parallel `GET` requests to the server cause multiple IPFS nodes to be created.  Better to have `res.locals.ipfs` be a promise to resolve an IPFS node as multiple contexts can await on it and it'll only resolve once.

On that `res.locals.ipfs` would be better as `req.app.ipfs` so the single IPFS node is explicitly global to the app rather than storing it on `res.locals` which imply per-request scoping.

Reply:
By which I mean, when I run this locally I see this sort of thing as requests from npm come in:

```console
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
😈 Spawning an in-process IPFS node using repo at /Users/alex/.jsipfs
(node:5539) UnhandledPromiseRejectionWarning: Error: Lock file is already being held
    at options.fs.stat (/Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/proper-lockfile/lib/lockfile.js:68:47)
    at /Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/graceful-fs/polyfills.js:285:20
    at FSReqWrap.oncomplete (fs.js:154:5)
(node:5539) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:5539) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:5539) UnhandledPromiseRejectionWarning: Error: Lock file is already being held
    at options.fs.stat (/Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/proper-lockfile/lib/lockfile.js:68:47)
    at /Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/graceful-fs/polyfills.js:285:20
    at FSReqWrap.oncomplete (fs.js:154:5)
(node:5539) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:5539) UnhandledPromiseRejectionWarning: Error: Lock file is already being held
    at options.fs.stat (/Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/proper-lockfile/lib/lockfile.js:68:47)
    at /Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/graceful-fs/polyfills.js:285:20
    at FSReqWrap.oncomplete (fs.js:154:5)
(node:5539) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
(node:5539) UnhandledPromiseRejectionWarning: Error: Lock file is already being held
    at options.fs.stat (/Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/proper-lockfile/lib/lockfile.js:68:47)
    at /Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/graceful-fs/polyfills.js:285:20
    at FSReqWrap.oncomplete (fs.js:154:5)
(node:5539) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 4)
(node:5539) UnhandledPromiseRejectionWarning: Error: Lock file is already being held
    at options.fs.stat (/Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/proper-lockfile/lib/lockfile.js:68:47)
    at /Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/graceful-fs/polyfills.js:285:20
    at FSReqWrap.oncomplete (fs.js:154:5)
(node:5539) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 5)
(node:5539) UnhandledPromiseRejectionWarning: Error: Lock file is already being held
    at options.fs.stat (/Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/proper-lockfile/lib/lockfile.js:68:47)
    at /Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/graceful-fs/polyfills.js:285:20
    at FSReqWrap.oncomplete (fs.js:154:5)
(node:5539) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 6)
(node:5539) UnhandledPromiseRejectionWarning: Error: Lock file is already being held
    at options.fs.stat (/Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/proper-lockfile/lib/lockfile.js:68:47)
    at /Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/graceful-fs/polyfills.js:285:20
    at FSReqWrap.oncomplete (fs.js:154:5)
(node:5539) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 7)
(node:5539) UnhandledPromiseRejectionWarning: Error: Lock file is already being held
    at options.fs.stat (/Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/proper-lockfile/lib/lockfile.js:68:47)
    at /Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/graceful-fs/polyfills.js:285:20
    at FSReqWrap.oncomplete (fs.js:154:5)
(node:5539) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 8)
(node:5539) UnhandledPromiseRejectionWarning: Error: Lock file is already being held
    at options.fs.stat (/Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/proper-lockfile/lib/lockfile.js:68:47)
    at /Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/graceful-fs/polyfills.js:285:20
    at FSReqWrap.oncomplete (fs.js:154:5)
(node:5539) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 9)
(node:5539) UnhandledPromiseRejectionWarning: Error: Lock file is already being held
    at options.fs.stat (/Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/proper-lockfile/lib/lockfile.js:68:47)
    at /Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/graceful-fs/polyfills.js:285:20
    at FSReqWrap.oncomplete (fs.js:154:5)
(node:5539) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 10)
(node:5539) UnhandledPromiseRejectionWarning: Error: Lock file is already being held
    at options.fs.stat (/Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/proper-lockfile/lib/lockfile.js:68:47)
    at /Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/graceful-fs/polyfills.js:285:20
    at FSReqWrap.oncomplete (fs.js:154:5)
(node:5539) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 11)
(node:5539) UnhandledPromiseRejectionWarning: Error: Lock file is already being held
    at options.fs.stat (/Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/proper-lockfile/lib/lockfile.js:68:47)
    at /Users/alex/Documents/Workspaces/ipfs-shipyard/npm-on-ipfs/node_modules/graceful-fs/polyfills.js:285:20
    at FSReqWrap.oncomplete (fs.js:154:5)
```

Reply:
Anyway, I fixed it in #111 but can't push to the incoming branch this PR is based on so have all the changes there.

Reply:
Hmm, that's weird, I don't think that happened to me 🤔

You're right, that is a better approach 👍